File: nomina/calculadora_prestaciones.py
"""Cálculo de provisiones de prestaciones sociales y bonificaciones institucionales."""

from __future__ import annotations

from datetime import date
from decimal import Decimal, ROUND_HALF_UP
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from modelo_datos import Contrato, Profesor
    from nomina.calculadora_deducciones import CalculadoraDeducciones


class CalculadoraPrestaciones:
    """Calcula cesantías, primas, vacaciones y bonificaciones especiales."""

    CERO = Decimal("0")
    DOS = Decimal("2")
    DIAS_ANIO = Decimal("360")

    def __init__(self, calculadora_deducciones: CalculadoraDeducciones) -> None:
        self.calc_ded = calculadora_deducciones

    @staticmethod
    def redondear(valor: Decimal) -> Decimal:
        return valor.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)

    def calcular_auxilio_transporte(
        self,
        contrato: Contrato,
        salario: Decimal,
        smmlv: Decimal,
        fecha: date | None = None,
        codigos_utilizados: dict[str, str] | None = None,
    ) -> Decimal:
        if contrato.aplicaAuxilioTransporte and salario <= self.DOS * smmlv:
            valor = self.calc_ded.obtener_parametro_decimal("VALOR_AUXILIO_TRANSPORTE_VIGENTE", fecha)
            if codigos_utilizados is not None:
                codigos_utilizados["VALOR_AUXILIO_TRANSPORTE_VIGENTE"] = str(valor) if valor is not None else str(self.CERO)
            return valor or self.CERO
        return self.CERO

    def calcular_bonificacion_posgrado(
        self,
        profesor: Profesor,
        smmlv: Decimal,
        contrato: Contrato,
        horas: Decimal | None,
        incluir_bonificaciones: bool,
        codigos_utilizados: dict[str, str] | None = None,
    ) -> Decimal:
        if contrato.permiteBonificacionPosgrado is False or not incluir_bonificaciones:
            return self.CERO
        factores = {
            "ESPECIALIZACION": Decimal("0.10"),
            "MAESTRIA": Decimal("0.45"),
            "DOCTORADO": Decimal("0.90"),
            "POSTDOCTORADO": self.CERO,
        }
        factor = factores.get(str(profesor.nivelPosgradoReconocido or "").upper(), self.CERO)
        if codigos_utilizados is not None:
            codigos_utilizados["BONIFICACION_POSGRADO"] = str(factor)
        return self._bonificacion_proporcional(smmlv * factor, contrato, horas)

    def calcular_bonificacion_investigacion(
        self,
        profesor: Profesor,
        smmlv: Decimal,
        contrato: Contrato,
        horas: Decimal | None,
        incluir_bonificaciones: bool,
        codigos_utilizados: dict[str, str] | None = None,
    ) -> Decimal:
        if contrato.permiteBonificacionInvestigacion is False or not incluir_bonificaciones:
            return self.CERO
        grupo = str(profesor.categoriaGrupoInvestigacion or "").upper()
        factores = {
            "GRUPO_A1": Decimal("0.56"), "A1": Decimal("0.56"),
            "GRUPO_A": Decimal("0.47"), "A": Decimal("0.47"),
            "GRUPO_B": Decimal("0.42"), "B": Decimal("0.42"),
            "GRUPO_C": Decimal("0.38"), "C": Decimal("0.38"),
            "GRUPO_RECONOCIDO": Decimal("0.33"), "SEMILLERO": Decimal("0.20"),
        }
        factor = factores.get(grupo, self.CERO)
        acreditada = bool(profesor.productividadInvestigativaVigente or profesor.participaProyectoInvestigacionVigente) and bool(profesor.certificacionVicerrectoriaInvestigacion)
        if codigos_utilizados is not None:
            codigos_utilizados["BONIFICACION_INVESTIGACION"] = str(factor) if factor else str(self.CERO)
        return self._bonificacion_proporcional(smmlv * factor if acreditada else self.CERO, contrato, horas)

    def _bonificacion_proporcional(self, valor: Decimal, contrato: Contrato, horas: Decimal | None) -> Decimal:
        if horas is None or contrato.horasMensualesAsignadas in (None, 0):
            return valor
        return valor * horas / Decimal(contrato.horasMensualesAsignadas)

    def calcular_provisiones(
        self,
        base_prestacional: Decimal,
        ibc: Decimal,
        dias: Decimal,
        *,
        regimen_especial: bool = False,
    ) -> dict[str, Decimal]:
        provisiones = {
            "cesantias": self.redondear(base_prestacional * dias / self.DIAS_ANIO),
            "intereses": self.redondear(base_prestacional * dias * Decimal("0.12") / self.DIAS_ANIO),
            "prima_servicios": self.redondear(base_prestacional * dias / self.DIAS_ANIO),
            "prima_navidad": self.redondear(base_prestacional * dias / self.DIAS_ANIO),
            "vacaciones": self.redondear(ibc * dias / Decimal("720")),
            "prima_vacaciones": self.CERO,
            "bonificacion_servicios": self.CERO,
        }
        if not regimen_especial:
            return provisiones

        # Bases especiales de los arts. 33, 39 y 46 del Decreto 1279
        tope = self.calc_ded.obtener_parametro_decimal("TOPE_BONIFICACION_SERVICIOS") or Decimal("756411")
        porcentaje = (
            self.calc_ded.obtener_porcentaje("PORCENTAJE_BONIFICACION_SERVICIOS_HASTA_TOPE", Decimal("0.50"))
            if ibc <= tope
            else self.calc_ded.obtener_porcentaje("PORCENTAJE_BONIFICACION_SERVICIOS_SOBRE_TOPE", Decimal("0.35"))
        )
        provisiones["bonificacion_servicios"] = self.redondear(ibc * porcentaje * dias / self.DIAS_ANIO)

        base_vacaciones = ibc + provisiones["prima_servicios"] / Decimal("12") + provisiones["bonificacion_servicios"] / Decimal("12")
        base_prima_vacaciones = (ibc * Decimal("2") / Decimal("3")) + provisiones["prima_servicios"] / Decimal("12") + provisiones["bonificacion_servicios"] / Decimal("12")

        provisiones["vacaciones"] = self.redondear(base_vacaciones * dias / Decimal("720"))
        provisiones["prima_vacaciones"] = self.redondear(base_prima_vacaciones * dias / Decimal("540"))

        base_prima_navidad = (
            ibc
            + provisiones["prima_servicios"] / Decimal("12")
            + provisiones["prima_vacaciones"] / Decimal("12")
            + provisiones["bonificacion_servicios"] / Decimal("12")
        )
        provisiones["prima_navidad"] = self.redondear(base_prima_navidad * dias / self.DIAS_ANIO)
        return provisiones

File: cpp/nomina/calculadora_prestaciones.cpp
#include "calculadora_prestaciones.h"
#include <algorithm>
#include <cmath>

namespace pita {

static std::string a_mayusculas(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c) { return std::toupper(c); });
    return s;
}

CalculadoraPrestaciones::CalculadoraPrestaciones(CalculadoraDeducciones& calculadoraDeducciones)
    : calcDed(calculadoraDeducciones) {}

double CalculadoraPrestaciones::redondear(double valor) {
    return std::round(valor * 100.0) / 100.0;
}

double CalculadoraPrestaciones::calcularAuxilioTransporte(
    const Contrato& contrato,
    double salario,
    double smmlv,
    const std::string& fecha,
    std::map<std::string, std::string>* codigosUtilizados
) {
    std::string dedicacion = contrato.dedicacion.has_value() ? to_string(*contrato.dedicacion) :
                            (contrato.tipoDedicacion.value_or(""));
    dedicacion = a_mayusculas(dedicacion);

    // Docentes de hora cátedra no devengan auxilio legal mensual salvo pacto
    if (dedicacion.find("CATEDRA") != std::string::npos && !contrato.aplicaAuxilioTransporte.value_or(false)) {
        return 0.0;
    }

    // Por mandato legal (Ley 15/1959), si devenga hasta 2 SMMLV aplica auxilio de transporte
    if (salario <= 2.0 * smmlv) {
        auto valorOpt = calcDed.obtenerParametroDecimal("VALOR_AUXILIO_TRANSPORTE_VIGENTE", fecha);
        if (codigosUtilizados) {
            (*codigosUtilizados)["VALOR_AUXILIO_TRANSPORTE_VIGENTE"] = valorOpt.has_value() ? std::to_string(*valorOpt) : "0";
        }
        return valorOpt.value_or(0.0);
    }
    return 0.0;
}

double CalculadoraPrestaciones::calcularBonificacionPosgrado(
    const Profesor& profesor,
    double smmlv,
    const Contrato& contrato,
    std::optional<double> horas,
    bool incluirBonificaciones,
    std::map<std::string, std::string>* codigosUtilizados
) {
    std::string tipoProf = profesor.tipoProfesor.has_value() ? to_string(*profesor.tipoProfesor) : "";
    std::string modContra = contrato.modalidadProfesor.value_or(contrato.tipoContrato.value_or(""));
    tipoProf = a_mayusculas(tipoProf);
    modContra = a_mayusculas(modContra);
    if (tipoProf.find("PLANTA") != std::string::npos || modContra.find("PLANTA") != std::string::npos) {
        return 0.0;
    }

    if ((contrato.permiteBonificacionPosgrado.has_value() && !*contrato.permiteBonificacionPosgrado) || !incluirBonificaciones) {
        return 0.0;
    }

    std::string nivel = profesor.nivelPosgradoReconocido.value_or(
        profesor.maximoNivelEstudio.value_or(
            contrato.nivelPosgradoAlVincular.value_or("")
        )
    );
    nivel = a_mayusculas(nivel);

    double factor = 0.0;
    if (nivel.find("DOCTOR") != std::string::npos) factor = 0.90;
    else if (nivel.find("MAESTR") != std::string::npos) factor = 0.45;
    else if (nivel.find("ESPEC") != std::string::npos) factor = 0.10;

    if (codigosUtilizados) {
        (*codigosUtilizados)["BONIFICACION_POSGRADO"] = std::to_string(factor);
    }
    return bonificacionProporcional(smmlv * factor, contrato, horas);
}

double CalculadoraPrestaciones::calcularBonificacionInvestigacion(
    const Profesor& profesor,
    double smmlv,
    const Contrato& contrato,
    std::optional<double> horas,
    bool incluirBonificaciones,
    std::map<std::string, std::string>* codigosUtilizados
) {
    if ((contrato.permiteBonificacionInvestigacion.has_value() && !*contrato.permiteBonificacionInvestigacion) || !incluirBonificaciones) {
        return 0.0;
    }
    std::string grupo = profesor.categoriaGrupoInvestigacion.has_value() ? a_mayusculas(*profesor.categoriaGrupoInvestigacion) : "";
    double factor = 0.0;
    if (grupo == "GRUPO_A1" || grupo == "A1") factor = 0.56;
    else if (grupo == "GRUPO_A" || grupo == "A") factor = 0.47;
    else if (grupo == "GRUPO_B" || grupo == "B") factor = 0.42;
    else if (grupo == "GRUPO_C" || grupo == "C") factor = 0.38;
    else if (grupo == "GRUPO_RECONOCIDO") factor = 0.33;
    else if (grupo == "SEMILLERO") factor = 0.20;

    bool prodVigente = profesor.productividadInvestigativaVigente.value_or(false) || profesor.participaProyectoInvestigacionVigente.value_or(false);
    bool certOk = profesor.certificacionVicerrectoriaInvestigacion.has_value() && !profesor.certificacionVicerrectoriaInvestigacion->empty();
    bool acreditada = (prodVigente && certOk);

    if (codigosUtilizados) {
        (*codigosUtilizados)["BONIFICACION_INVESTIGACION"] = (factor > 0.0) ? std::to_string(factor) : "0";
    }
    return bonificacionProporcional(acreditada ? (smmlv * factor) : 0.0, contrato, horas);
}

double CalculadoraPrestaciones::bonificacionProporcional(double valor, const Contrato& contrato, std::optional<double> horas) {
    if (!horas.has_value() || !contrato.horasMensualesAsignadas.has_value() || *contrato.horasMensualesAsignadas == 0.0) {
        return (valor >= 100.0) ? std::round(valor) : valor;
    }
    double res = valor * (*horas) / (*contrato.horasMensualesAsignadas);
    return (res >= 100.0) ? std::round(res) : res;
}

std::map<std::string, double> CalculadoraPrestaciones::calcularProvisiones(
    double basePrestacional,
    double ibc,
    double dias,
    bool regimenEspecial
) {
    std::map<std::string, double> provisiones;
    double cesantias = redondear(basePrestacional * dias / 360.0);
    double intereses = redondear(cesantias * dias * 0.12 / 360.0);
    provisiones["cesantias"] = cesantias;
    provisiones["intereses"] = intereses;
    provisiones["prima_servicios"] = redondear(basePrestacional * dias / 360.0);
    provisiones["prima_navidad"] = regimenEspecial ? redondear(basePrestacional * dias / 360.0) : 0.0;
    provisiones["vacaciones"] = redondear(ibc * dias / 720.0);
    provisiones["prima_vacaciones"] = 0.0;
    provisiones["bonificacion_servicios"] = 0.0;

    if (!regimenEspecial) {
        return provisiones;
    }

    double tope = calcDed.obtenerParametroDecimal("TOPE_BONIFICACION_SERVICIOS").value_or(756411.0);
    double porcentaje = (ibc <= tope)
        ? calcDed.obtenerPorcentaje("PORCENTAJE_BONIFICACION_SERVICIOS_HASTA_TOPE", 0.50)
        : calcDed.obtenerPorcentaje("PORCENTAJE_BONIFICACION_SERVICIOS_SOBRE_TOPE", 0.35);

    provisiones["bonificacion_servicios"] = redondear(ibc * porcentaje * dias / 360.0);

    double baseVacaciones = ibc + (provisiones["prima_servicios"] / 12.0) + (provisiones["bonificacion_servicios"] / 12.0);
    double basePrimaVacaciones = (ibc * 2.0 / 3.0) + (provisiones["prima_servicios"] / 12.0) + (provisiones["bonificacion_servicios"] / 12.0);

    provisiones["vacaciones"] = redondear(baseVacaciones * dias / 720.0);
    provisiones["prima_vacaciones"] = redondear(basePrimaVacaciones * dias / 540.0);

    double basePrimaNavidad = ibc + (provisiones["prima_servicios"] / 12.0)
                                + (provisiones["prima_vacaciones"] / 12.0)
                                + (provisiones["bonificacion_servicios"] / 12.0);

    provisiones["prima_navidad"] = redondear(basePrimaNavidad * dias / 360.0);
    return provisiones;
}

} // namespace pita
